File: backend/internal/extraction/prompts.go
package extraction

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"text/template"
)

const DefaultPromptSetDir = "prompts/task_processing_pipeline_v1"

type PromptSet struct {
	Version           string
	Dir               string
	htmlFromImage     *template.Template
	parameters        *template.Template
	generation        *template.Template
	selfEvaluation    *template.Template
	defaultSourceHTML string
}

type promptSetManifest struct {
	Version string         `json:"version"`
	Files   promptSetFiles `json:"files"`
}

type promptSetFiles struct {
	HTMLFromImage  string `json:"html_from_image"`
	Parameters     string `json:"parameters"`
	Generation     string `json:"generation"`
	SelfEvaluation string `json:"self_evaluation"`
	DefaultSource  string `json:"default_source"`
}

type generationPromptData struct {
	TaskHTML       string
	Params         string
	UserComment    string
	HasUserComment bool
}

func LoadPromptSet(dir string) (*PromptSet, error) {
	dir = strings.TrimSpace(dir)
	if dir == "" {
		dir = DefaultPromptSetDir
	}

	manifestPath := filepath.Join(dir, "manifest.json")
	data, err := os.ReadFile(manifestPath)
	if err != nil {
		return nil, fmt.Errorf("read prompt manifest %q: %w", manifestPath, err)
	}

	var manifest promptSetManifest
	if err := json.Unmarshal(data, &manifest); err != nil {
		return nil, fmt.Errorf("parse prompt manifest %q: %w", manifestPath, err)
	}
	if strings.TrimSpace(manifest.Version) == "" {
		return nil, fmt.Errorf("prompt manifest %q has empty version", manifestPath)
	}
	if err := manifest.Files.validate(); err != nil {
		return nil, fmt.Errorf("prompt manifest %q: %w", manifestPath, err)
	}

	defaultSourceHTML, err := readPromptFile(dir, manifest.Files.DefaultSource)
	if err != nil {
		return nil, err
	}
	htmlFromImage, err := parsePromptTemplate(dir, "html_from_image", manifest.Files.HTMLFromImage)
	if err != nil {
		return nil, err
	}
	parameters, err := parsePromptTemplate(dir, "parameters", manifest.Files.Parameters)
	if err != nil {
		return nil, err
	}
	generation, err := parsePromptTemplate(dir, "generation", manifest.Files.Generation)
	if err != nil {
		return nil, err
	}
	selfEvaluation, err := parsePromptTemplate(dir, "self_evaluation", manifest.Files.SelfEvaluation)
	if err != nil {
		return nil, err
	}

	return &PromptSet{
		Version:           strings.TrimSpace(manifest.Version),
		Dir:               dir,
		htmlFromImage:     htmlFromImage,
		parameters:        parameters,
		generation:        generation,
		selfEvaluation:    selfEvaluation,
		defaultSourceHTML: defaultSourceHTML,
	}, nil
}

func (f promptSetFiles) validate() error {
	if strings.TrimSpace(f.HTMLFromImage) == "" {
		return fmt.Errorf("files.html_from_image is required")
	}
	if strings.TrimSpace(f.Parameters) == "" {
		return fmt.Errorf("files.parameters is required")
	}
	if strings.TrimSpace(f.Generation) == "" {
		return fmt.Errorf("files.generation is required")
	}
	if strings.TrimSpace(f.SelfEvaluation) == "" {
		return fmt.Errorf("files.self_evaluation is required")
	}
	if strings.TrimSpace(f.DefaultSource) == "" {
		return fmt.Errorf("files.default_source is required")
	}
	return nil
}

func parsePromptTemplate(dir, name, filename string) (*template.Template, error) {
	content, err := readPromptFile(dir, filename)
	if err != nil {
		return nil, err
	}
	tpl, err := template.New(name).Option("missingkey=error").Parse(content)
	if err != nil {
		return nil, fmt.Errorf("parse prompt template %q: %w", filepath.Join(dir, filename), err)
	}
	return tpl, nil
}

func readPromptFile(dir, filename string) (string, error) {
	filename = strings.TrimSpace(filename)
	if filename == "" {
		return "", fmt.Errorf("prompt filename is empty")
	}
	path := filepath.Join(dir, filename)
	data, err := os.ReadFile(path)
	if err != nil {
		return "", fmt.Errorf("read prompt file %q: %w", path, err)
	}
	return strings.TrimRight(string(data), "\n"), nil
}

func (p *PromptSet) HTMLFromImagePrompt() (string, error) {
	return executePromptTemplate(p.htmlFromImage, nil)
}

func (p *PromptSet) ParametersPrompt(taskHTML string) (string, error) {
	return executePromptTemplate(p.parameters, map[string]string{"TaskHTML": taskHTML})
}

func (p *PromptSet) GenerationPrompt(taskHTML string, params taskParameters, userComment string) (string, error) {
	userComment = strings.TrimSpace(userComment)
	return executePromptTemplate(p.generation, generationPromptData{
		TaskHTML:       taskHTML,
		Params:         formatTaskParameters(params),
		UserComment:    userComment,
		HasUserComment: userComment != "",
	})
}

func (p *PromptSet) SelfEvaluationPrompt(sourceHTML, variantHTML string) (string, error) {
	return executePromptTemplate(p.selfEvaluation, map[string]string{
		"SourceHTML":  sourceHTML,
		"VariantHTML": variantHTML,
	})
}

func (p *PromptSet) DefaultSourceHTML() string {
	return p.defaultSourceHTML
}

func executePromptTemplate(tpl *template.Template, data any) (string, error) {
	var buf bytes.Buffer
	if err := tpl.Execute(&buf, data); err != nil {
		return "", err
	}
	return buf.String(), nil
}

File: variantor-be/app/pipeline/prompts.py
import json
import re
from pathlib import Path


GO_PLACEHOLDER_RE = re.compile(r"{{\s*\.([A-Za-z0-9_]+)\s*}}")
GO_IF_RE = re.compile(r"{{\s*if\s+\.([A-Za-z0-9_]+)\s*}}(.*?){{\s*end\s*}}", re.DOTALL)


class PromptSet:
    def __init__(self, directory: str):
        self.directory = Path(directory)
        manifest = json.loads((self.directory / "manifest.json").read_text(encoding="utf-8"))
        self.version = manifest["version"]
        files = manifest["files"]
        self._html_from_image = self._read(files["html_from_image"])
        self._parameters = self._read(files["parameters"])
        self._generation = self._read(files["generation"])
        self._self_evaluation = self._read(files["self_evaluation"])
        self.default_source = self._read(files["default_source"])

    def html_from_image_prompt(self) -> str:
        return self._render(self._html_from_image, {})

    def parameters_prompt(self, task_html: str) -> str:
        return self._render(self._parameters, {"TaskHTML": task_html})

    def generation_prompt(self, task_html: str, params: dict, user_comment: str = "") -> str:
        clean_comment = user_comment.strip()
        return self._render(
            self._generation,
            {
                "TaskHTML": task_html,
                "Params": format_task_parameters(params),
                "UserComment": clean_comment,
                "HasUserComment": bool(clean_comment),
            },
        )

    def self_evaluation_prompt(self, source_html: str, variant_html: str) -> str:
        return self._render(self._self_evaluation, {"SourceHTML": source_html, "VariantHTML": variant_html})

    def _read(self, filename: str) -> str:
        return (self.directory / filename).read_text(encoding="utf-8").rstrip("\n")

    def _render(self, template: str, values: dict) -> str:
        def replace_if(match: re.Match) -> str:
            key = match.group(1)
            body = match.group(2)
            return body if values.get(key) else ""

        rendered = GO_IF_RE.sub(replace_if, template)
        rendered = GO_PLACEHOLDER_RE.sub(lambda match: str(values.get(match.group(1), "")), rendered)
        return rendered


def format_task_parameters(params: dict) -> str:
    return "\n".join(
        [
            "Тип задания: " + value_or_star(params.get("task_type")),
            "Предполагаемый класс: " + value_or_star(params.get("school_class")),
            "Уровень сложности задания: " + value_or_star(params.get("difficulty")),
        ]
    )


def value_or_star(value: object) -> str:
    text = str(value or "").strip()
    return text or "*"
